### PR TITLE
Refactor caching, including use of msgpack for serialization

### PR DESCRIPTION
**Pull request by oxtopus**
This is a superset of https://github.com/numenta/cortipy/pull/35 and introduces some speedups.

### CODE REVIEW

**Comment by oxtopus** [comment]
Looks like I need to fixup some unit tests.  Hold of on merge until travis is green, but please review the changes in the meantime.

File: cortipy/cortical_client.py [hunk 1]
```diff
@@ -134,13 +83,66 @@ def __init__(self,
...
     self.cacheDir = cacheDir
     self.verbosity = verbosity
     self.retina = retina
-    self.useCache = useCache
+    self._session = requests.Session()
+
+
+  def _cachedRequest(self, fn, url, params, headers, data=None):
+    def _doRequest():
+      # Internal request wrapper to
```
Comment (subutai) [inline cortipy/cortical_client.py]: Dangling sentence :smile:

File: cortipy/cortical_client.py [hunk 2]
```diff
@@ -134,13 +83,66 @@ def __init__(self,
...
+
+    self.useCache = useCache
     self.cacheDir = cacheDir
     self.verbosity = verbosity
     self.retina = retina
-    self.useCache = useCache
+    self._session = requests.Session()
+
+
+  def _cachedRequest(self, fn, url, params, headers, data=None):
```
Comment (subutai) [inline cortipy/cortical_client.py:96]: I think a docstring would be useful here.

File: cortipy/cortical_client.py [hunk 3]
```diff
@@ -134,13 +83,66 @@ def __init__(self,
...
+    cacheKey = m.hexdigest()
+
+    cachePath = os.path.join(self.cacheDir, "{}.msgpack".format(cacheKey))
+
+    if os.path.isfile(cachePath):
+      # Request has been made before, load and return original response
+      response = msgpack.load(open(cachePath, "r"))
+    else:
+      # Make new request, cache the response
+      response = _doRequest()
```
Comment (subutai) [inline cortipy/cortical_client.py:153]: What happens if CIO API was unable to fulfill this request or if they return an empty bitmap? Shouldn't we generate a placeholder bitmap here?

**Comment by subutai** [comment]
I think these are good changes, so :+1: I believe it still leaves in the existing bug where the placeholder encoding for an un-encodable term is not cached properly.  We could fix that in a subsequence PR.

File: cortipy/cortical_client.py (hunk 3 above)
Comment (oxtopus) [inline cortipy/cortical_client.py:153]: The cache decorator implementation that this replaces didn't have that functionality, so I didn't add it here.  I figure that will be addressed as part of RES-183, independent of these changes.  The primary optimization in this changeset is the use of `msgpack` as opposed to `json` for serialization.

**Comment by oxtopus** [comment]
Thanks, @subutai.  I added the docstring and will merge once tests are fixed.

File: cortipy/cortical_client.py (hunk 3 above)
Comment (subutai) [inline cortipy/cortical_client.py:153]: OK, makes sense.

**Comment by BoltzmannBrain** [comment]
:+1:

**Pull request merged**